src/ser.rs

```diff
@@ -19,449 +19,204 @@ use serde::ser;
 use super::error::{Error, Result};
 use private;
 
-pub struct Serializer;
-
-impl ser::Serializer for Serializer {
-    type Ok = Yaml;
-    type Error = Error;
-
-    type SerializeSeq = SerializeArray;
-    type SerializeTuple = SerializeArray;
-    type SerializeTupleStruct = SerializeArray;
-    type SerializeTupleVariant = SerializeTupleVariant;
-    type SerializeMap = SerializeMap;
-    type SerializeStruct = SerializeStruct;
-    type SerializeStructVariant = SerializeStructVariant;
-
-    fn serialize_bool(self, v: bool) -> Result<Yaml> {
-        Ok(Yaml::Boolean(v))
-    }
-
-    fn serialize_i8(self, v: i8) -> Result<Yaml> {
-        self.serialize_i64(v as i64)
-    }
-
-    fn serialize_i16(self, v: i16) -> Result<Yaml> {
-        self.serialize_i64(v as i64)
-    }
-
-    fn serialize_i32(self, v: i32) -> Result<Yaml> {
-        self.serialize_i64(v as i64)
-    }
-
-    fn serialize_i64(self, v: i64) -> Result<Yaml> {
-        Ok(Yaml::Integer(v))
-    }
-
-    serde_if_integer128! {
-        #[cfg_attr(feature = "cargo-clippy", allow(cast_possible_truncation))]
-        fn serialize_i128(self, v: i128) -> Result<Yaml> {
-            if v <= i64::max_value() as i128 && v >= i64::min_value() as i128 {
-                self.serialize_i64(v as i64)
-            } else {
-                Ok(Yaml::Real(v.to_string()))
-            }
-        }
-    }
+/// A structure for serializing Rust values into YAML.
+pub struct Serializer<W> {
+    writer: W,
+}
 
-    fn serialize_u8(self, v: u8) -> Result<Yaml> {
-        self.serialize_i64(v as i64)
+impl<W> Serializer<W>
+where
+    W: io::Write,
+{
+    /// Creates a new YAML serializer.
+    #[inline]
+    pub fn new(writer: W) -> Serializer<W> {
+        Serializer { writer }
     }
 
-    fn serialize_u16(self, v: u16) -> Result<Yaml> {
-        self.serialize_i64(v as i64)
+    /// Unwrap the `Writer` from the `Serializer`.
+    pub fn into_inner(self) -> W {
+        self.writer
     }
+}
 
-    fn serialize_u32(self, v: u32) -> Result<Yaml> {
-        self.serialize_i64(v as i64)
-    }
+impl<'a, W> ser::Serializer for &'a mut Serializer<W>
+where
+    W: io::Write,
+{
+    type Ok = ();
+    type Error = Error;
 
-    fn serialize_u64(self, v: u64) -> Result<Yaml> {
-        if v <= i64::max_value() as u64 {
-            self.serialize_i64(v as i64)
+    type SerializeSeq = &'a mut Serializer<W>;
+    type SerializeTuple = &'a mut Serializer<W>;
+    type SerializeTupleStruct = &'a mut Serializer<W>;
+    type SerializeTupleVariant = &'a mut Serializer<W>;
+    type SerializeMap = &'a mut Serializer<W>;
+    type SerializeStruct = &'a mut Serializer<W>;
+    type SerializeStructVariant = &'a mut Serializer<W>;
+
+    #[inline]
+    fn serialize_bool(self, value: bool) -> Result<()> {
```

Comment:
@dtolnay I wonder if this is the correct method to do this or we should patch it to https://github.com/chyh1990/yaml-rust/issues/127?